Teach_Breakout.py

```diff
@@ -0,0 +1,344 @@
+        is_terminals.append(mem_is_terminal.recall(index))
+
+    # We should now have a full batch, which the DNN can train on
+    fit_batch_target(model, target_model, gamma, start_states, actions, rewards, next_states, is_terminals, number_in_batch)
+
+
+def q_iteration(env, model, target_model, iteration, mem_frames, mem_actions, mem_rewards, mem_is_terminal, mem_size):
+    """
+    Do one iteration of acting then learning
+    """
+
+    # Choose epsilon based on the iteration
+    epsilon = get_epsilon_for_iteration(iteration)
+
+    start_state = make_state(mem_frames, iteration)
+    # Choose the action:
+    if random.random() < epsilon:
+        action = env.action_space.sample()
+    else:
+        action = choose_best_action(model, start_state)
+    # Play one game iteration (note: according to a paper, you should actually play 4 times here)
+    new_frame, reward, is_terminal, _ = env.step(action)
+
+    reward = transform_reward(reward)  # Make reward just +1 or -1
+    new_frame = preprocess(new_frame)  # Preprocess frame before saving it
+
+    # Add to memory
+    mem_frames.append(new_frame, iteration+1)  # The frame we ended up in after the action
+    mem_actions.append(action, iteration+1)  # The action we did
+    mem_rewards.append(reward, iteration+1)  # The reward we received after doing the action
+    mem_is_terminal.append(is_terminal, iteration+1)  # Whether or not the new frame is terminal
+
+    # Make then fit a batch (gamma=0.99, num_in_batch=?)
+    number_in_batch = 32
+    make_n_fit_batch(model, target_model, 0.99, iteration, mem_size, mem_frames, mem_actions, mem_rewards, mem_is_terminal, number_in_batch)
+
+    # If DONE, reset model!
+    if is_terminal:
+        env.reset()
+
+    return action, reward, is_terminal, epsilon
+
+#----------------------------------------------------------------------------------------------------------------------#
+
+def main():
+    """
+    Train the DNN to play Breakout
+    """
+    import argparse
+    parser = argparse.ArgumentParser()
+    parser.add_argument('-r', '--num_rand_acts', help="Random actions before learning starts",
+                        default=5*10**4, type=int)
+    parser.add_argument('-s', '--save_after', help="Save after this number of training steps",
+                        default=10**5, type=int)
+    parser.add_argument('-m', '--mem_size', help="Size of the experience replay memory",
+                        default=10**6, type=int)
+    parser.add_argument('-sn', '--save_name', help="Name of the saved models", default=None, type=str)
+    #parser.add_argument('-g', '--greedy_after', default=1e6, type=int)
+    args = parser.parse_args()
+
+
+    global greedy_after
+    global start_at
+
+    # Other things to modify
+    greedy_after = 10**6    # 1e6 in the paper
+    start_at = 1          # 1 in the paper
+    number_training_steps = 10**8 # It should start doing well after 1e6??
+    Print_progress_after = 10 ** 2
```

Comment:
Python variables should be `written_in_lowercase`, except constants that are `ALL_CAPS`